### ISSUE TITLE
golang 1.15 cannot connect to docker socket API due to Transfer-encoding issue

### ISSUE THREAD

**Issue by h8liu** [docker/for-mac#4855]
- [x] I have tried with the latest version of my channel (Stable or Edge)
  - [x] I have uploaded Diagnostics
  - Diagnostics ID: 6422B13B-4FBA-4F67-B6D5-E7E5A4792C7D/20200821053405

### Expected behavior

Expect golang http client can use all docker APIs via the docker UDS.
For more details see https://github.com/golang/go/issues/40735

### Actual behavior

Returns error on doing POST:
net/http: HTTP/1.x transport connection broken: unsupported transfer encoding: "identity" 

### Information

  - macOS Version: Catalina Version 10.15.5
  - Is it reproducible? Yes.
  - Is the problem new? Yes.
  - Did the problem appear with an update? No.

### Diagnostic logs

(I don't think diagnostic logs matter here.)

```
Docker for Mac: version...
$ docker version
Client: Docker Engine - Community
 Version:           19.03.12
 API version:       1.40
 Go version:        go1.13.10
 Git commit:        48a66213fe
 Built:             Mon Jun 22 15:41:33 2020
 OS/Arch:           darwin/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          19.03.12
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.13.10
  Git commit:       48a66213fe
  Built:            Mon Jun 22 15:49:27 2020
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          v1.2.13
  GitCommit:        7ad184331fa3e55e52b890ea95e65ba581ae3429
 runc:
  Version:          1.0.0-rc10
  GitCommit:        dc9208a3303feef5b3839f4323d9beb36df0a9dd
 docker-init:
  Version:          0.18.0
  GitCommit:        fec3683
```

### Steps to reproduce the behavior

See https://github.com/golang/go/issues/40735 for how to reproduce.

tl;dr

1. `docker create --name test alpine sleep 1000`
2. `docker start test`
3. `go1.15 run a.go`, where a.go is:

```
package main

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io/ioutil"
	"log"
	"net"
	"net/http"
	"net/url"
	"strings"
)

func main() {
	// Before running this needs to:
	// - Have docker 19.03.12 installed
	// - Run "docker create --name test alpine sleep 10000" to create the container
	// - Run "docker start test" to start it.
	if err := run(); err != nil {
		log.Fatal(err)
	}
}

func run() error {
	d := new(net.Dialer)
	x := &http.Transport{
		DialContext: func(ctx context.Context, net, addr string) (net.Conn, error) {
			return d.DialContext(ctx, "unix", "/var/run/docker.sock")
		},
	}

	c := &http.Client{Transport: x}

	u := &url.URL{
		Scheme: "http",
		Host:   "uds",
		Path:   "/v1.40/containers/test/exec",
	}
	var req = struct {
		Cmd          []string
		AttachStdout bool
		AttachStderr bool
	}{
		Cmd:          []string{"echo", "hello"},
		AttachStderr: true,
		AttachStdout: true,
	}
	body, err := json.Marshal(req)
	if err != nil {
		return err
	}
	header := make(http.Header)
	header.Add("Content-Type", "application/json")
	httpReq := &http.Request{
		Method: "POST",
		URL:    u,
		Header: header,
		Body:   ioutil.NopCloser(bytes.NewBuffer(body)),
	}
	resp, err := c.Do(httpReq)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	bodyBytes, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return err
	}
	var createResp struct {
		ID string `json:"Id"`
	}
	if err := json.Unmarshal(bodyBytes, &createResp); err != nil {
		return err
	}

	u2 := &url.URL{
		Scheme: "http",
		Host:   "uds",
		Path:   "/v1.40/exec/" + createResp.ID + "/start",
	}

	httpReq2 := &http.Request{
		Method: "POST",
		URL:    u2,
		Header: header,
		Body:   ioutil.NopCloser(strings.NewReader("{}")),
	}
	resp2, err := c.Do(httpReq2)
	if err != nil {
		return err
	}

	defer resp2.Body.Close()
	bodyBytes2, err := ioutil.ReadAll(resp2.Body)
	if err != nil {
		return err
	}
	fmt.Println(string(bodyBytes2))

	return nil
}
```

**Comment by h8liu** [comment]
The same issue does not reproduce on Linux (with very close docker versions that are compiled with the same golang version), so I think it is OSX specific, due to some mishandling on the unix domain socket proxying between the docker server (which runs in a Linux VM-ish thing?) and the client.

**Comment by stephen-turner** [comment]
The go ticket says that this has broken between go 1.14 and 1.15, so something must have changed in go.

**Comment by djs55** [comment]
I believe this problem has been fixed in the latest Docker Desktop Edge 2.3.7.0 (48173): https://docs.docker.com/docker-for-mac/edge-release-notes/#docker-desktop-community-2370 
```
$ go run a.go
hello

$ go version
go version go1.15.2 darwin/amd64
```
Let us know how you get on. I'll close this ticket now but feel free to open new tickets if you encounter other problems.

Thanks for the detailed report and repro case -- it's much appreciated!
